File: src/largestRectangleInHistogram.py
from typing import List


class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        maxArea = 0
        stack = []
        for i, h in enumerate(heights):
            start = i
            while stack and stack[-1][1] > h:
                ind, height = stack.pop()
                maxArea = max(maxArea, (i-ind)*height)
                start = ind
                '''
                there is problem I ran into. how do I keep track of the point
                from where that particular area began? like, if we had 6 in stack 
                and new addition was 5, then we will pop. the area, however, will be
                5 units from the 6 and 5 units from the 5th one. how do we know that 
                there are 5 units from the the 6th block also. this is where the start 
                variable comes into play. using this variable, we can keep track of the
                earliest(i.e, the first block from which it got an area) block.
                '''
            stack.append((start,h))


        '''
        after completing above portion, our stack will be a monotonically decreasing stack
        or just an empty stack. the below for loop is used to deal with the case of a non
        empty stack.  
        '''
        for i, h in stack:
            maxArea = max(maxArea, h * (len(heights) - i))
        return maxArea

File: src/largestRectangleInHistogram.java
import java.util.Stack;

class Solution {
    class Tuple{
        private int ind;
        private int h;
        Tuple(int ind ,int h){
            this.ind = ind;
            this.h = h;
        }

        public int getInd() {
            return ind;
        }

        public int getH() {
            return h;
        }

    }
    public int largestRectangleArea(int[] heights) {
        int maxArea = 0;
        Stack<Tuple> stack = new Stack<>();
        stack.push(new Tuple(0, heights[0]));
        for(int i=1; i < heights.length; i++){
            int start = i;
            while (!stack.isEmpty() && stack.peek().getH() > heights[i]){
                Tuple x = stack.pop();
                int h = x.getH();
                int ind = x.getInd();
                maxArea = Integer.max(maxArea, (i-ind)*h);
                start = ind;
            }
            stack.push(new Tuple(start, heights[i]));
        }
        for (Tuple x : stack) {
            int h = x.getH();
            int ind = x.getInd();
            maxArea = Integer.max(maxArea, (heights.length - ind) * h);
        }
        return maxArea;
    }
}

/*
how to solve this problem?
1) create a monotonically increasing stack. which means that the stack will always be in increasing order.
2) if current element is less than stack top, then keep popping the stack until the stack top is less than current element.
3) in this popping part, we will use two vars, height and index. height is the height of that histogram and index is basically the point(probably behind)
from where we include in the map.
4) we are using a start var to keep track of the index from where we are starting to include in the map.
5) last fop loop is to calculate the area os stack elements which survived the popping part.
*/